modules/proxy/proxy_util.c

```diff
@@ -2620,6 +2697,248 @@ PROXY_DECLARE(int) ap_proxy_release_connection(const char *proxy_function,
+                     * address to expire at some point (for connect() issues).
+                     */
+                    /* TODO: use a monotonic clock here */
+                    /* Recompute now should the DNS be slow */
+                    now = apr_time_sec(apr_time_now() - *proxy_start_time);
+                    address->expiry = apr_atomic_read32(&worker->s->address_expiry);
+                    if (address->expiry <= now) {
+                        apr_uint32_t next_expiry = address->expiry + worker->s->address_ttl;
+                        while (next_expiry <= now) {
+                            next_expiry += worker->s->address_ttl;
+                        }
+                        next_expiry = apr_atomic_cas32(&worker->s->address_expiry,
+                                                       next_expiry, address->expiry);
+                        if (next_expiry != address->expiry) {
+                            /* Race lost, well the expiry should grow anyway.. */
+                            AP_DEBUG_ASSERT(next_expiry > now);
+                            address->expiry = next_expiry;
+                        }
+                    }
+                }
+                else {
+                    address->expiry = APR_UINT32_MAX;
+                }
+                if (!old_address) {
+                    /* First address (i.e. worker->cp->addr) never expires */
+                    apr_atomic_set32(&address->refcount, 2);
+                }
+                else {
+                    apr_atomic_set32(&address->refcount, 1);
+                }
+
+                /* Acquire the new address for the connection with the lifetime
+                 * of its pool, first releasing the previous one eventually.
+                 */
+                if (conn->address) {
+                    apr_pool_cleanup_run(conn->pool, conn->address,
+                                         proxy_address_cleanup);
+                }
+                conn->address = address;
+                proxy_address_inc(address);
+                apr_pool_cleanup_register(conn->pool, address,
+                                          proxy_address_cleanup,
+                                          apr_pool_cleanup_null);
+
+                /* Publish the changes. The old worker address (if any) is no
+                 * longer used by this worker, it can be destroyed now if the
+                 * worker is the last user (refcount == 1) or by the last conn
+                 * using it (refcount > 1).
+                 */
+                AP_VOLATILIZE_T(proxy_address *, worker->address) = address;
+                if (old_address) {
+                    proxy_address_dec(old_address);
+                }
+
+                address_changed = 1;
+            }
+
+            /* Update worker->cp->addr if it's not set (or reset) */
+            if (!AP_VOLATILIZE_T(apr_sockaddr_t *, worker->cp->addr)) {
+                worker->cp->addr = address->addr;
```

Comment:
How do we ensure it never expires even if it got reset in the meantime?

Reply:
The first address/DNS lookup will never expire because it has a refcount of 2, but if some (third party) module plays by resetting worker->cp->addr arbitrarily after the startup it will be set to the current address, which will expire when a new one is needed and the worker or last connection using it releases it.
I don't think it's ever been safe to reset worker->cp->addr, so we should be good?

Reply:
I agree that using setting `worker->cp->addr` to `NULL` as an indicator was not a good idea and given that this possibility is probably only available in one patch release of 2.4. I am personally happy to drop this. I only feel unsure if we could do it for formal reasons. But probably as long as no one complains we get away with it :-p.